File: Talleres RESFTUL JAVA DOTNET/Dotnet/API-DOTNET-BANK/Controllers/TransactionController.cs
using API_DOTNET_BANK.Application.DTOs.Transaction;
using API_DOTNET_BANK.Application.Interface;
using API_DOTNET_BANK.Application.Service;
using API_DOTNET_BANK.Infraestructure;
using API_DOTNET_BANK.Infraestructure.Repository;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace API_DOTNET_BANK.Web.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class TransactionController : ControllerBase
    {
        private readonly AppDbContext _dbContext;

        public TransactionController(AppDbContext dbContext)
        {
            _dbContext = dbContext;
        }

        private ITransactionService CreateService()
        {
            var transactionRepository = new TransactionRepository(_dbContext);
            var accountRepository = new AccountRepository(_dbContext);
            return new TransactionService(transactionRepository, accountRepository);
        }

        [HttpGet]
        public async Task<IActionResult> GetAll()
        {
            var transactions = await CreateService().GetAllAsync();
            return Ok(transactions);
        }

        [HttpGet("{id}")]
        public async Task<IActionResult> GetById(int id)
        {
            var transaction = await CreateService().GetByIdAsync(id);
            if (transaction == null)
                return NotFound("Transacción no encontrada");
            return Ok(transaction);
        }

        [HttpGet("account/{accountId}")]
        public async Task<IActionResult> GetByAccountId(int accountId)
        {
            var transactions = await CreateService().GetByAccountIdAsync(accountId);
            return Ok(transactions);
        }

        [HttpPost("deposit")]
        public async Task<IActionResult> Deposit([FromBody] CreateDepositDto dto)
        {
            var transaction = await CreateService().DepositAsync(dto);
            if (transaction == null)
                return BadRequest("No se pudo realizar el depósito. Verifique que la cuenta exista y esté activa.");
            return CreatedAtAction(nameof(GetById), new { id = transaction.Id }, transaction);
        }

        [HttpPost("withdrawal")]
        public async Task<IActionResult> Withdraw([FromBody] CreateWithdrawalDto dto)
        {
            var transaction = await CreateService().WithdrawAsync(dto);
            if (transaction == null)
                return BadRequest("No se pudo realizar el retiro. Verifique que la cuenta exista, esté activa y tenga saldo suficiente.");
            return CreatedAtAction(nameof(GetById), new { id = transaction.Id }, transaction);
        }

        [HttpPost("transfer")]
        public async Task<IActionResult> Transfer([FromBody] CreateTransferDto dto)
        {
            var transaction = await CreateService().TransferAsync(dto);
            if (transaction == null)
                return BadRequest("No se pudo realizar la transferencia. Verifique que ambas cuentas existan, estén activas y la cuenta origen tenga saldo suficiente.");
            return CreatedAtAction(nameof(GetById), new { id = transaction.Id }, transaction);
        }
    }
}

File: Talleres RESFTUL JAVA DOTNET/Java/api-java-bank/src/main/java/com/allpasoft/apijavabank/web/controller/TransactionController.java
package com.allpasoft.apijavabank.web.controller;

import com.allpasoft.apijavabank.application.dto.*;
import com.allpasoft.apijavabank.application.service.TransactionService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.net.URI;
import java.util.List;

@RestController
@RequestMapping("/api/transaction")
@RequiredArgsConstructor
@Tag(name = "Transacciones", description = "Endpoints para depósitos, retiros y transferencias")
public class TransactionController {

    private final TransactionService transactionService;

    @GetMapping
    @Operation(
        summary = "Listar todas las transacciones",
        description = "Obtiene la lista de todas las transacciones",
        responses = {
            @ApiResponse(responseCode = "200", description = "Lista de transacciones obtenida exitosamente")
        }
    )
    public ResponseEntity<List<TransactionDto>> getAll() {
        return ResponseEntity.ok(transactionService.getAll());
    }

    @GetMapping("/{id}")
    @Operation(
        summary = "Obtener transacción por ID",
        description = "Obtiene una transacción por su ID",
        responses = {
            @ApiResponse(responseCode = "200", description = "Transacción obtenida exitosamente"),
            @ApiResponse(responseCode = "404", description = "Transacción no encontrada")
        }
    )
    public ResponseEntity<Object> getById(@PathVariable Long id) {
        return transactionService.getById(id)
                .<ResponseEntity<Object>>map(ResponseEntity::ok)
                .orElse(ResponseEntity.status(404).body("Transacción no encontrada"));
    }

    @GetMapping("/account/{accountId}")
    @Operation(
        summary = "Obtener transacciones por cuenta",
        description = "Obtiene todas las transacciones de una cuenta",
        responses = {
            @ApiResponse(responseCode = "200", description = "Transacciones obtenidas exitosamente")
        }
    )
    public ResponseEntity<List<TransactionDto>> getByAccountId(@PathVariable Long accountId) {
        return ResponseEntity.ok(transactionService.getByAccountId(accountId));
    }

    @PostMapping("/deposit")
    @Operation(
        summary = "Realizar depósito",
        description = "Realiza un depósito en una cuenta",
        responses = {
            @ApiResponse(responseCode = "201", description = "Depósito realizado exitosamente"),
            @ApiResponse(responseCode = "400", description = "No se pudo realizar el depósito")
        }
    )
    public ResponseEntity<Object> deposit(@RequestBody CreateDepositDto dto) {
        return transactionService.deposit(dto)
                .<ResponseEntity<Object>>map(t -> ResponseEntity.created(URI.create("/api/transaction/" + t.getId())).body(t))
                .orElse(ResponseEntity.badRequest().body("No se pudo realizar el depósito. Verifique que la cuenta exista y esté activa."));
    }

    @PostMapping("/withdrawal")
    @Operation(
        summary = "Realizar retiro",
        description = "Realiza un retiro de una cuenta",
        responses = {
            @ApiResponse(responseCode = "201", description = "Retiro realizado exitosamente"),
            @ApiResponse(responseCode = "400", description = "No se pudo realizar el retiro")
        }
    )
    public ResponseEntity<Object> withdraw(@RequestBody CreateWithdrawalDto dto) {
        return transactionService.withdraw(dto)
                .<ResponseEntity<Object>>map(t -> ResponseEntity.created(URI.create("/api/transaction/" + t.getId())).body(t))
                .orElse(ResponseEntity.badRequest().body("No se pudo realizar el retiro. Verifique que la cuenta exista, esté activa y tenga saldo suficiente."));
    }

    @PostMapping("/transfer")
    @Operation(
        summary = "Realizar transferencia",
        description = "Realiza una transferencia entre cuentas",
        responses = {
            @ApiResponse(responseCode = "201", description = "Transferencia realizada exitosamente"),
            @ApiResponse(responseCode = "400", description = "No se pudo realizar la transferencia")
        }
    )
    public ResponseEntity<Object> transfer(@RequestBody CreateTransferDto dto) {
        return transactionService.transfer(dto)
                .<ResponseEntity<Object>>map(t -> ResponseEntity.created(URI.create("/api/transaction/" + t.getId())).body(t))
                .orElse(ResponseEntity.badRequest().body("No se pudo realizar la transferencia. Verifique que ambas cuentas existan, estén activas y la cuenta origen tenga saldo suficiente."));
    }
}
